includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -1199,7 +1199,20 @@ public function process_payment_for_order( $cart, $payment_information, $additio
 				throw new Exception( WC_Payments_Utils::get_filtered_error_message( $e ) );
 			}
 
-			$selected_payment_method_type = [ WC_Payments::get_gateway()->get_selected_stripe_payment_type_id() ];
+			// If UPE is enabled, get the selected payment method.
+			if ( ! WC_Payments_Features::is_upe_enabled() ) {
+				$payment_methods = WC_Payments::get_gateway()->get_payment_method_ids_enabled_at_checkout( null, true );
+			} else {
+				$upe_payment_method = sanitize_text_field( wp_unslash( $_POST['payment_method'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification
+
+				if ( 'woocommerce_payments' !== $upe_payment_method ) {
+					$upe_payment_method = str_replace( 'woocommerce_payments_', '', $upe_payment_method );
+				} else {
+					$upe_payment_method = 'card';
+				}
+
+				$payment_methods = [ $upe_payment_method ];
+			}
```

Comment:
```suggestion

			$payment_gateway = sanitize_text_field( wp_unslash( $_POST['payment_method'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification

			if ( 'woocommerce_payments' !== $payment_gateway ) {
				$payment_gateway = str_replace( 'woocommerce_payments_', '', $payment_gateway );
			} else {
				$payment_gateway = 'card';
			}

			$payment_methods = [ $payment_gateway ];

```
The logic here looks good to me...buuut unfortunately this entire `WC_Payments_Features::is_upe_enabled()` is soon going to be removed in #4919, so we won't be able to use that as a condition here. Though, wouldn't that second block work in either case, whether the gateway is a UPE gateway or not? Can we not just use that on its own?